pennylane/labs/resource_estimation/resource_contianer.py

```diff
@@ -0,0 +1,111 @@
+# Copyright 2024 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+r"""Base classes for the resource objects and resource estimation."""
+from collections import defaultdict
+from dataclasses import dataclass, field
+
+
+class OpTypeWithParams:
+
+    def __init__(self, op_type: str, params_tuple: tuple) -> None:
```

Comment:
I think the types could be a bit more descriptive. Instead of `op_type: str` we should pass in a callable that produces a PennyLane Operator from `params_tuple`. I'm thinking something like this `x = OpTypeWithParams(qml.QFT, (('wires', [0, 1]))` and the corresponding type annotation should look like `op_type: Callable[[tuple], Operator]`. With this change we can still access the string through the `__name__ method, but now we can do error handling by checking `issubclass(op_type, Operator)` to ensure that the instantiated object corresponds to a PennyLane Operator.

Reply:
Ahh no, we don't want to do this. Imagine that we have an operation that acts on lots of wires, we would rather store `num_wires` then explicitly a list of all of the wire ids that are required to instantiate the operator. 

This is the difference between the parameters required for full decomposition vs parameters required for resource level decompositions.

Reply:
We could potentially pass around references to the class itself. I don't know how well that plays with things like hashing. Since its easier to hash a tuple of strings and ints, then it is to hash classes in memory

Reply:
I think we should pass in the class itself as input, that way we can check to make sure it is an Operator. Then we can hash on the string `op_type.__name__`.

Reply:
Added 👍🏼